### PR TITLE
ACPI battery support

### PR DESCRIPTION
**Pull request by no92**

### CODE REVIEW

**Comment by no92** [comment]
Reworked the patches to use a shared kernel fiber, and rebased on master.

**Comment by d-tatianin** [comment]
LGTM as far as ACPI and uACPI usage

**Review by Dennisbonke — APPROVED** [review]
LGTM

**Review by qookei — APPROVED** [review]

File: kernel/thor/system/acpi/madt.cpp [hunk 1]
```diff
@@ -370,4 +373,23 @@ static initgraph::Task initPmInterfaceTask{&globalInitEngine, "acpi.init-pm-inte
...
+		Scheduler::resume(acpiFiber);
+	}
+};
+
+static initgraph::Task initBatteriesTask{&globalInitEngine, "acpi.init-batteries",
+	initgraph::Requires{&loadAcpiNamespaceTask, &initAcpiWorkqueueTask},
+	[] {
+		initializeBatteries();
+	}
+};
```
Comment (avdgrinten) [inline kernel/thor/system/acpi/madt.cpp:393]: Why is this in `madt.cpp`?

File: protocols/hw/src/client.cpp [hunk 2]
```diff
@@ -568,5 +568,57 @@ async::result<helix::UniqueDescriptor> Device::accessFbMemory() {
 	co_return std::move(bar);
 }
 
+#define COPY_BATTERY_STATE(x) do { \
+	if(resp.x()) \
+		state.x = resp.x(); \
+	else \
+		state.x = std::nullopt; \
+} while(0)
```
Comment (avdgrinten) [inline protocols/hw/src/client.cpp:576]: This is an unnecessary hack.

Comment (no92) [reply protocols/hw/src/client.cpp]: Replaced this with a lambda.

**Review by avdgrinten — APPROVED** [review]

**Pull request merged**